### PR TITLE
[SPARK-42573][Connect][Scala] Enable binary compatibility tests on all major client APIs

### PR DESCRIPTION
**Pull request by zhenlineo**
### What changes were proposed in this pull request?
Make binary compatibility check for SparkSession/Dataset/Column/functions etc.

### Why are the changes needed?
Help us to have a good understanding of the current API coverage of the Scala client.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests.

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 1]
```diff
@@ -69,30 +69,131 @@ class CompatibilitySuite extends AnyFunSuite { // scalastyle:ignore funsuite
...
-      ProblemFilters.exclude[Problem]("org.apache.spark.connect.proto.*"),
-      // Disable Range until we support typed APIs
+
+      // Dataset
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.ofRows"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.DATASET_ID_TAG"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.COL_POS_KEY"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.DATASET_ID_KEY"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.curId"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.groupBy"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:101]: This should be binary compatible.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 2]
```diff
@@ -69,30 +69,131 @@ class CompatibilitySuite extends AnyFunSuite { // scalastyle:ignore funsuite
...
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.encoder"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.sqlContext"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.as"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.checkpoint"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.localCheckpoint"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.withWatermark"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.na"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.stat"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.joinWith"),
+      ProblemFilters.exclude[Problem]("org.apache.spark.sql.Dataset.select"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:113]: This should be binary compatible.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala [hunk 3]
```diff
@@ -69,30 +69,131 @@ class CompatibilitySuite extends AnyFunSuite { // scalastyle:ignore funsuite
...
-      // IncludeByName("org.apache.spark.sql.Dataset$"),
-      IncludeByName("org.apache.spark.sql.DataFrame"),
+      IncludeByName("org.apache.spark.sql.Column.*"),
+      IncludeByName("org.apache.spark.sql.DataFrame.*"),
       IncludeByName("org.apache.spark.sql.DataFrameReader.*"),
       IncludeByName("org.apache.spark.sql.DataFrameWriter.*"),
       IncludeByName("org.apache.spark.sql.DataFrameWriterV2.*"),
-      IncludeByName("org.apache.spark.sql.SparkSession"),
-      IncludeByName("org.apache.spark.sql.SparkSession$")) ++ includeImplementedMethods(clientJar)
+      IncludeByName("org.apache.spark.sql.Dataset.*"),
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:77]: Can you also add Column?

Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:77]: The column is fully compatible. See line 72. Already added.

Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:77]: Hmmm, `Column.as` is missing in the connect client.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 1 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:101]:
```
method groupBy(java.lang.String,Array[java.lang.String])org.apache.spark.sql.RelationalGroupedDataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
method groupBy(java.lang.String,scala.collection.Seq)org.apache.spark.sql.RelationalGroupedDataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
```
Seems indeed missing these two. If you commit it out, you will get the above message.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 2 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:113]:
```
[info]   method select(org.apache.spark.sql.TypedColumn)org.apache.spark.sql.Dataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
[info]   method select(org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn)org.apache.spark.sql.Dataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
[info]   method select(org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn)org.apache.spark.sql.Dataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
[info]   method select(org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn)org.apache.spark.sql.Dataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
[info]   method select(org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn,org.apache.spark.sql.TypedColumn)org.apache.spark.sql.Dataset in class org.apache.spark.sql.Dataset does not have a correspondent in client version
```

We do not have `org.apache.spark.sql.TypedColumn` yet

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 3 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:77]: No way, we have all 5 `Colum.as` :D

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 2 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:113]: Yeah you are right

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/CompatibilitySuite.scala:101]: Yeah we are missing a signature.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by hvanhovell** [comment]
Merging, thanks for doing this!

**Pull request closed without merging**